plugins/plugin-machine/che-plugin-machine-ext-server/src/main/java/org/eclipse/che/ide/ext/machine/server/ssh/KeysInjector.java

```diff
@@ -29,97 +45,130 @@
  */
 @Singleton // must be eager
 public class KeysInjector {
+
   private static final Logger LOG = LoggerFactory.getLogger(KeysInjector.class);
 
   private final EventService eventService;
-  private final DockerConnector docker;
   private final SshManager sshManager;
-  // TODO replace with WorkspaceManager
-  //    private final CheEnvironmentEngine environmentEngine;
+  private final WorkspaceManager workspaceManager;
+  private final ExecAgentClientFactory execAgentClientFactory;
 
   @Inject
   public KeysInjector(
-      EventService eventService, DockerConnector dockerConnector, SshManager sshManager
-      //                        CheEnvironmentEngine environmentEngine
-      ) {
+      EventService eventService,
+      SshManager sshManager,
+      WorkspaceManager workspaceManager,
+      ExecAgentClientFactory execAgentClientFactory) {
     this.eventService = eventService;
-    this.docker = dockerConnector;
     this.sshManager = sshManager;
-    //        this.environmentEngine = environmentEngine;
+    this.workspaceManager = workspaceManager;
+    this.execAgentClientFactory = execAgentClientFactory;
   }
 
   @PostConstruct
   public void start() {
     eventService.subscribe(
-        new EventSubscriber<MachineStatusEvent>() {
+        new EventSubscriber<WorkspaceStatusEvent>() {
```

Comment:
This subscriber looks quite heavy. I think it would be better to not hold event service thread such a long time on such a common event as workspace running event.